expanded weights without fast rules

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70141
* __->__ #70140

[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights
 - The rest is the implementation of the erroring fallback + the mechanism for being able to register faster per sample grad rules. Only linear is implemented here, but they are all implemented in #70141

Differential Revision: [D34350950](https://our.internmc.facebook.com/intern/diff/D34350950)

Commit:
expanded weights without fast rules

[ghstack-poisoned]

samdow commented on torch/expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,188 @@
+        result.append(None)
+        return tuple(result)
+
+# Fallback:
+# - forward pass: run operation with de-expanded input
+# - backward pass: run autograd.grad in a for-loop to compute per-sample-grads.
+#                  This is NOT something the vmap API can handle, something more
+#                  low-level is at work here.
+
+
+# We can override the fallback by implementing efficient "per-sample-grad" rules
+# for the backward pass. The forward pass should still be the same, though.
```
move this comment above the class

samdow commented on torch/expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,188 @@
+    def __repr__(self):
+        return "ExpandedWeight for:\n" + self.orig_weight.__repr__() + f" with batch size {self.batch_size}"
+
+class MyFunction(torch.autograd.Function):
```
rename class to be more descriptive

samdow commented on torch/expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,188 @@
+        return None
+
+    @property
+    def dtype(self):
```
move near shape so all grads are together

samdow commented on torch/expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,188 @@
+        return autograd_func
+    return decorator
+
+def forward_helper(func, ctx, expanded_args, num_true_outs):
```
comment

Commit:
Update on "expanded weights without fast rules"




[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"




[ghstack-poisoned]

zou3519 commented on test/test_expanded_weights.py:
```diff
@@ -0,0 +1,121 @@
+
```
nit: empty line

zou3519 commented on torch/nn/utils/_stateless.py:
```diff
@@ -43,3 +44,39 @@ def functional_call(module, parameters_and_buffers, args, kwargs=None):
         else:
             out = module(args, **kwargs)
     return out
+
+def per_sample_call(module, batch_size, args, kwargs=None):
```
Location seems fine for now (since we're not ready to go beta in pytorch 1.11) but long term this should probably go into torch.nn.utils?

samdow replied:
Moved to `_per_sample_grad.py` just so it doesn't get too messy in `_stateless` and added a comment to that function--basically because this depends on `_stateless.function_call` which imports `torch.nn.Module` we get a circular dependency if we try to expose this in `torch.nn.utils`

zou3519 commented on torch/nn/utils/_stateless.py:
```diff
@@ -43,3 +44,39 @@ def functional_call(module, parameters_and_buffers, args, kwargs=None):
+    Invoked just like a forward pass, ``per_sample_call`` will produce the same
+    forward result. Then, when backward is invoked, the parameters of ``module``
+    will have a ``grad_sample`` field populated with the per sample gradients
+    instead of the batched gradients
```
nit: instead of the regular gradients?

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,203 @@
+            kwargs = {key: cls.conv_kwarg_defaults[key] for key in remaining_kwargs_options} | kwargs
+        return cls.handled_functions[func].apply(*(args + tuple(kwargs.values())))
+
+    @property
+    def shape(self):
+        return self.orig_weight.shape
```
What happens if you call `expanded_weight.size()`? does that return the correct thing?

samdow replied:
Added `size()` function

zou3519 replied:
Follow-up question: (maybe more for @albanD) is user error to pass a `__torch_function__` tensor subclass into some C++ code that requires a Tensor? For the PyTorch frontend API the answer to this is probably no because it should enter `__torch_function__` but if users have things like custom C++ operators that they've pybind'ed into Python, do we consider this to be a user error?

albanD replied:
It is not an error and the code will use the underlying Tensor as-is. This is how `nn.Parameter` works today.
But it is a subtlety that the user needs to be aware, as soon as you pass the frontend API binding or enter any other c++ function. Only the c++ Tensor associated with your subclass will exist.

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,203 @@
+        if func not in cls.handled_functions:
+            warnings.warn(f"don't have custom implementation for function {func.__name__}. Using slow fallback")
+            return SlowFallback.apply(func, *(args + tuple(kwargs.values())))
+        if func == torch.nn.functional.conv2d:
+            remaining_kwargs = 7 - len(args)
+            remaining_kwargs_options = cls.conv_kwarg_options[4 - remaining_kwargs:]
+            kwargs = {key: cls.conv_kwarg_defaults[key] for key in remaining_kwargs_options} | kwargs
```
Hmm, do we need to do this default arg thing for all functions that have default arguments?

samdow replied:
So this was confusing to say the least because other functions have default arguments (and have tests that don't pass all of those default arguments) but none of them were running into the problem that conv2d was. I'm not sure if this will be fixed by the convolution refactor. My best guess was that the `handle_torch_function` call was not passing some default arguments because [others are always passing them](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2375-L2387) but it was a little hard to verify this in the generated code.

zou3519 replied:
Seems like a general problem: https://github.com/pytorch/pytorch/issues/45565

zou3519 replied:
If there are not too many of these (e.g. just conv2d) -- we could use the hack that you're doing here. We should add some parity testing though just in case the defaults ever change.

iirc torch.fx had a similar problem and they wrote something like their own "default argument filler" - maybe Horace or James would know more but they're out at the moment.

zou3519 replied:
Resolved offline: OpInfos have the case with defaults if we trust them.

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,203 @@
+    output = func(*unexpanded_args)
+    return _check_and_detach_output(ctx, output, num_true_outs)
+
+def _check_and_unexpand_args(func, ctx, expanded_args):
```
Does the expanded weights fallback correctly raise an error on operations where we cannot detect if the tensor has a bdim or not? (or is the behavior here that we hope the batch dim is in the correct place?) For example, if the user does something like:

```
torch.add(input, expanded_weight)
```
We probably want to block that operation from happening because we don't know where the batch dim is in the input! It could be in dimension 0, it could be in dimension 1, it might not have one.

samdow replied:
Currently it doesn't! I think you're right that we should be erroring because of that case but not sure if that means that we should eliminate the fallback entirely then and just error if we can't deal with it. I think we only know the batch dimension location for the `nn.functional` operations, so we could try to look for those operators. Even then, we're assuming that their batch dimension is in dimension 0 which might not be true for all of them

zou3519 replied:
I see, maybe we should eliminate the fallback entirely and then just error if we can't deal with it

samdow replied:
Fallback eliminated and design doc updated

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,203 @@
+        expanded_args = func_and_expanded_args[1:]
+        with torch.enable_grad():
+            output = forward_helper(func, ctx, expanded_args, 1)
+            true_output = ctx.true_outputs
```
nit: you can `del ctx.true_outputs` after this, right?

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,203 @@
+                result.append(torch.stack(cur_per_sample_grad).sum(0))
+            else:
+                result.append(None)
+        result.append(None)
```
I'm a bit confused about all these Nones, but maybe a comment with an example will help clear that up

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,203 @@
+            if not isinstance(true_output, torch.Tensor):
+                raise RuntimeError(f"Fallback only works on Tensor output, got output was {type(true_output).__name__}")
+            outputs = tuple(true_output[i] for i in range(true_output.shape[0]))
+        ctx.outputs = outputs
```
cc @albanD are there problems with this? I remember you saying something about memory leaks

It looks like we've detached the output already, so maybe this is OK

samdow replied:
Resolving since fallback is eliminated

zou3519 reviewed (commented):
Need to run but here are some initial comments

zou3519 commented on test/test_expanded_weights.py:
```diff
@@ -0,0 +1,121 @@
+from torch.testing._internal.common_nn import TestBase, module_tests, new_module_tests
+
+from torch.overrides import handle_torch_function, has_torch_function_variadic
+
```
Some more suggestions for tests:
- create an ExpandedWeight object and check that all of its fields (shape, dtype, etc) work.
- Test error cases: If the user does something weird with an ExpandedWeight we should try to error out gracefully. E.g. if the user tries to do torch.add(regular_tensor, ExpandedWeight) then that is probably a problem because we can't infer where the batch dim on regular_tensor is.

samdow replied:
Added (1) as `TestExpandedWeightAttributes` and switched the fallback case to be a testing error case

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,203 @@
+# - backward pass: run autograd.grad in a for-loop to compute per-sample-grads.
+#                  This is NOT something the vmap API can handle, something more
+#                  low-level is at work here.
+class SlowFallback(torch.autograd.Function):
```
does this work for in-place operations? Like the torch.nn.init functions?

I'm not sure how likely it is for people to be using those. People will definitely use torch.nn.functional.relu w/ in-place though

samdow replied:
Resolving since fallback is eliminated

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,92 @@
+        ret = torch.Tensor._make_subclass(cls, orig_weight.detach(), orig_weight.requires_grad)
+        if not isinstance(orig_weight, torch.Tensor):
+            raise RuntimeError(f"Can only make Expanded Weights of Tensors, got {type(orig_weight).__name__}")
+        ret.batch_size = batch_size
+        ret.orig_weight = orig_weight
+        return ret
```
mypy _really_ does not like this because "Tensor does not have attribute `batch_size`". Any thoughts? Otherwise I'll keep playing around with it and see if we can make something work

zou3519 replied:
Does a typing.cast() work here? (e.g. cast `ret` to ExpandedWeight. This is a no-op for everything aside from mypy.

Otherwise a mypy ignore (I forget the syntax) here is fine

zou3519 commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,92 @@
+    def shape(self):
+        return self.orig_weight.shape
+
+    def size(self):
+        return self.orig_weight.size()
```
Other things you might need to override:
- `tensor.size(1)`
- `tensor.numel()`
- `tensor.stride()`
- `tensor.is_contiguous()`

zou3519 replied:
I'm a little confused though -- since we did `ret = torch.Tensor._make_subclass(cls, orig_weight.detach(), orig_weight.requires_grad)`, then I would expect all of these to already work and that we don't need to provide overrides for them

samdow replied:
It looks like all of them are going through `__torch_function__`. To your other comment, I was able to find [this](https://github.com/HEmile/storchastic/blob/master/storch/tensor.py) that uses `__torch_function__` and it looks like there's a lot I still need to override (though a couple, like `getattr` work fine now so I'll check as I'm implementing/testing)

zou3519 commented on docs/source/nn.rst:
```diff
@@ -416,6 +416,15 @@ for more information on how to implement your own parametrizations.
 
     parametrize.ParametrizationList
 
+Stateless API
```
Is this the right heading?

zou3519 commented on torch/nn/utils/_per_sample_grad.py:
```diff
@@ -0,0 +1,40 @@
+
+# dependency on `functional_call` means that this can't be exposed in utils
+# without creating circular dependency
+def per_sample_call(module, batch_size, args, kwargs=None):
```
nit: we might want to bikeshed this name some more. I would be really confused if I saw `per_sample_call(model)` in someone's code if I didn't already know about per-sample-gradients

samdow replied:
Yeah definitely see that--what about `call_with_per_sample_gradients`?

zou3519 commented on torch/nn/utils/_per_sample_grad.py:
```diff
@@ -0,0 +1,40 @@
+        >>> assert model.bias.grad_sample.shape == (5, 3)
+        >>> assert model.bias.grad == None
+    """
+    def maybe_build_expanded_weight(og_tensor):
+        if og_tensor.requires_grad:
+            return ExpandedWeight(og_tensor, batch_size)
+        else:
+            return og_tensor
+
+    params = {name: maybe_build_expanded_weight(value) for (name, value) in module.named_parameters()}
```
functional_call's second argument is `params_and_buffers`, right? We don't want to create an expanded weight object on anything that does not require gradient in that case.

Also, are there any nn.Modules that use buffers? (Aside from BatchNorm). That should be a test case somewhere

samdow replied:
[Instance norm](https://pytorch.org/docs/stable/generated/torch.nn.InstanceNorm1d.html#torch.nn.InstanceNorm1d) has this since it extends from batch norm's base and the functional version accepts running_mean/running_var

zou3519 commented on torch/nn/utils/_per_sample_grad.py:
```diff
@@ -0,0 +1,40 @@
+        >>> assert model.bias.grad_sample.shape == (5, 3)
+        >>> assert model.bias.grad == None
+    """
+    def maybe_build_expanded_weight(og_tensor):
+        if og_tensor.requires_grad:
+            return ExpandedWeight(og_tensor, batch_size)
+        else:
+            return og_tensor
```
you probably want to do some validation on the batch_size (check it is >= 0? Is 0 valid?) and `module` (check isinstance(module, torch.nn.Module))

zou3519 commented on torch/_expanded_weights/expanded_weights_utils.py:
```diff
@@ -0,0 +1,116 @@
+import torch
+from .expanded_weights_impl import ExpandedWeight
+
```
Nothing in this file gets used anymore now that we've deleted the fallback, right? Maybe adding a single simple per-sample-grad rule to this PR (like the one for linear) would help demonstrate how these are used, especially for testing

zou3519 replied:
(Or just moving this file to the next PR or somewhere else)

zou3519 reviewed (commented):
The code in this PR looks good to me but I have a suggestion around organization and testing. This PR introduced:
1. The ExpandedWeights Object
2. The per_sample_grad API
3. A lot of helper functions

Only (1) is being tested in test_expanded_weights.py. (2) and (3) are probably tested in the next PR in the stack (I haven't looked yet) but in general each PR in a stack should be able to stand by themselves. Maybe we should include the per-sample-grad rule for simple layer (like linear) in this PR so we can test the per_sample_call API as well as the helper functions and demonstrate how everything works end-to-end.

Regarding the ExpandedWeights Object -- is there a list of common Tensor attributes to override somewhere? I think we are missing `stride()`, `is_contiguous()`, `memory_format()`, but there might be more. It might be good to find some prominent user of `__torch_function__` (I don't know of any) and see what attributes they override

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented on torch/nn/utils/_per_sample_grad.py:
```diff
@@ -0,0 +1,47 @@
+    Examples::
+        >>> model = nn.Linear(4, 3)
+        >>> batched_input = torch.randn(5, 4)  # batch size of 5
+        >>> res = per_sample_call(model, batched_input.shape[0], batched_input).sum()
```
change to `call_for_per_sample_grads`

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

jbschlosser commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,146 @@
+    def __torch_function__(cls, func, _, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func == torch.nn.functional.conv2d:
```
Reminder to also support `conv1d` / `conv3d`

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented:
In this update:
 - Added conv1d/3d casing (support added in next PR)
 - Removed all unnecessary methods + attributes (decided offline)
 - Removed public documentation

Still TODO:
 - move from `torch/_expanded_weights` to `torch/nn/utils/_expanded_weights`

samdow commented on torch/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,78 @@
+        # i.e. torch.add(torch.Tensor, ExpandedWeight)
+        raise RuntimeError(f"Expanded Weights encountered but cannot handle function {func.__name__}")
+
+    @property
+    def dtype(self):
+        return self.orig_weight.dtype
+
+    @property
+    def shape(self):
+        return self.orig_weight.shape
```
These two are necessary for reparameterization and, per our offline discussion, we actually need the shape to be that of the unexpanded tensor for reparamaterization to be allowed

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented:
@zou3519 @albanD @jbschlosser This is ready for re-review if any of you have remaining comments

zou3519 commented on torch/nn/utils/_expanded_weights/linear_expanded_weights.py:
```diff
@@ -0,0 +1,23 @@
+        (input, weight, bias) = ctx.args
+        results = []
+
+        results.append(grad_if_exists_for_input(input, lambda: grad_output.matmul(unpack_expanded_weight_or_tensor(weight))))
```
Hmm, we're really baking in the assumption that the ExpandedWeight is wrapping a leaf tensor (a tensor that requires grad that is a leaf of the computation graph). We probably want a more generalized version of ExpandedWeight in the future, where users can e.g. call .relu() on an ExpandedWeight object to return a new ExpandedWeight object.

How much do you think we should worry about what we need in the future? I'm not sure -- I think Opacus bakes in this assumption as well? If we wanted to test our hypothesis that "this assumption works for most models" then we should do something like demonstrate running ExpandedWeights over the torchbenchmark suite of models. OTOH we're not exposing this yet, so another approach would be to merge this first "as a prototype" and then validate next.

zou3519 reviewed (commented):
Some questions, but I had two high-level comments:
1. We're really baking in the assumption that ExpandedWeights wrap leaf tensors. Is this OK?
2. It's a bit difficult to follow the logic of what gets stored on the context in the forward pass and why and how they get used further down the line. The LinearPerSampleGrad for example doesn't seem to need to store the output, but that appears to happen in `forward_helper`?

Some suggestions for (2) could be:
- It might be clearer for `forward_helper` to be a pure function that returns a bunch of tensors that LinearPerSampleGrad.forward then decides to save to ctx or not. Ditto for grad_if_exists_for_input.
- More documentation (e.g. forward_helper makes no mention of modifying ctx); but in general making the code clearer lessens the need for documentation

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented on torch/nn/utils/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,78 @@
+            ordered_kwargs = tuple(kwargs.get(key, cls.conv_kwarg_defaults[key]) for key in remaining_kwargs_options)
+            return cls.handled_functions[func].apply(*(args + ordered_kwargs))
+        if func in cls.handled_functions:
+            return cls.handled_functions[func].apply(*(args + tuple(kwargs.values())))
```
The order is determined by insertion order but because this runs through `__torch_function__` handling, that order is fixed by the generated code (i.e. users listing the kwargs in different order won't change this). The one thing with changing it to take kwargs instead is that we'll have to have some special case handling for every function like we have with conv because the arguments can either be positional args or kwargs

samdow replied:
Thinking about this more: we actually would be able to take kwargs for the same reason: we know that `__torch_function__` will put these kwargs in a dict, even if the user passed them positonally.

However, when I try to do this I get: `TypeError: apply() takes no keyword arguments` so we need to do it positionally

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -613,6 +613,8 @@ def __init__(self,
                  test_neg_view=True,
                  assert_jit_shape_analysis=False,  # assert that jit shape analysis fully propagates shape
+                 # the following metadata relates to ExpandedWeights support and is checked in test_ops.py
+                 supports_expanded_weight=False,
```
Fixed comment

samdow commented on torch/nn/utils/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,78 @@
+            a = a.orig_weight
+        if isinstance(b, ExpandedWeight):
+            b = b.orig_weight
+        return torch.allclose(a, b, rtol, atol, equal_nan)
```
This was from a now removed test. Thanks for the catch! Removed `allclose` support

samdow commented on torch/nn/utils/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,78 @@
+            remaining_kwargs = 7 - len(args)
+            remaining_kwargs_options = cls.conv_kwarg_options[4 - remaining_kwargs:]
+            ordered_kwargs = tuple(kwargs.get(key, cls.conv_kwarg_defaults[key]) for key in remaining_kwargs_options)
+            return cls.handled_functions[func].apply(*(args + ordered_kwargs))
```
Moved to next PR

samdow commented on torch/nn/utils/_expanded_weights/expanded_weights_utils.py:
```diff
@@ -0,0 +1,122 @@
+    elif num_true_outs != 1:
+        raise RuntimeError(f"Got single output but expected at least {num_true_outs} outputs. "
+                           "Issues in ExpandedWeights' autograd.Function")
+    ctx.true_outputs = output
```
Fixed--moved handling to the function level (i.e. group_norm, instance_norm, and layer_norm now unwrap the tuple of a single element) so that we could support (Tensor,) return in the future

samdow commented on torch/nn/utils/_expanded_weights/expanded_weights_utils.py:
```diff
@@ -0,0 +1,114 @@
+    unpacked = unpack_expanded_weight_or_tensor(maybe_expanded_weight)
+    if isinstance(maybe_expanded_weight, ExpandedWeight):
+        if hasattr(unpacked, "grad_sample"):
+            if isinstance(unpacked.grad_sample, list):
```
This will put the grad_samples together in a list if there's more than one. This follows what Opacus does and has some nice flexibilit. Typically you will want the stacked gradients (this is now done in a single stack versus having to do multiple stacks if we returned a tensor, not a list), but in the case of RNNs, users will want the per sample gradients to be accumulated for RNN cell (Opacus has a special layer to do this). We would still be able to support accumulation behavior for RNNs, the summing would just be taken care of by Opacus instead of autograd

Results still showed significant speedup for ExpandedWeights over Opacus, even though we're saving more memory

samdow replied:
Also added a couple small tests that repeats the tests checking the per sample gradients, but is called twice with the same inputs. The test checks that the behavior matched what would happen if we had a single batch with the inputs repeated twice

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented:
@zou3519 The remaining test failures look unrelated to the PR.  I can try to look at CI and rebase if it's green.

Added in this update:
- Functions take in the flattened args and the keys for all kwargs so we can reconstruct the kwargs in the backend
- Linear's torch_function was moved to the generated layer which means that, like conv, we have to add back in any default kwargs that weren't passed. Moved this to the function level too instead of being in `expanded_weights_impl`
- Multiple per sample gradient computations with the same underlying weight will cause it to be accumulated because this is what happens in RNNs. There's also an error if we try to run `call_with_per_sample_grads` while any of the weights have the `grad_sample` field set

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

zou3519 commented on test/test_expanded_weights.py:
```diff
@@ -0,0 +1,409 @@
+        module = self.constructor(*self.constructor_args)
+        input = self._get_input()
+        if len(input.shape) == 0 or input.shape[0] == 0:
+            return
```
might be able to return "skipped"

samdow replied:
Fixed

zou3519 commented on test/test_expanded_weights.py:
```diff
@@ -0,0 +1,409 @@
+            return
+        test_case._do_test_multi_input(module, input)
+
+# TODO: Once all of these use ModuleInfo, replace with ModuleInfo tests
```
Comment more on this is using the legacy nn tests... and that we'll move them eventully

samdow replied:
Fixed

zou3519 commented on torch/nn/utils/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,57 @@
+class ExpandedWeight(torch.Tensor):
+    def __init__(self, orig_weight, batch_size):
+        self.batch_size = batch_size
+        self.orig_weight = orig_weight
```
The ExpandedWeight object has orig_weight's data, and then we're assigning orig_weight to it. So this effectively doubles the parameter

zou3519 commented on torch/nn/utils/_expanded_weights/expanded_weights_impl.py:
```diff
@@ -0,0 +1,57 @@
+#
+# This is a __torch_function__ object but it could have also been a Tensor Extension
+# with a dispatch key.
+class ExpandedWeight(torch.Tensor):
```
Good to note this is torch.Tensor b/c/ reparameterization

zou3519 commented on torch/nn/utils/_expanded_weights/linear_expanded_weights.py:
```diff
@@ -0,0 +1,35 @@
+        results = []
+
+        results.append(grad_if_exists_for_input(input, lambda: grad_output.matmul(unpack_expanded_weight_or_tensor(weight))))
+        results.extend([None] * 3)  # weight and bias don't compute batched gradients
```
input, weight, bias, kwarg_names are the inputs

samdow replied:
The update hopefully made this easier to understand because kwarg_names is at the front so the None for it gets added at the start instead of here

zou3519 commented on torch/nn/utils/_expanded_weights/expanded_weights_utils.py:
```diff
@@ -0,0 +1,124 @@
+                         for (name, arg) in expanded_kwargs.items()}
+    return unexpanded_args, unexpanded_kwargs
+
+def _check_and_detach_output(output, num_true_outs):
```
TODO(rzou): read the next PR up in the stack to see -- I'm getting by num_true_outs and aux_output but it makes sense

zou3519 reviewed (commented):
TODO:
- check to see if _make_wrapper_subclass works
- (rzou): aux_output, num_true_outputs is a bit weird, check next stack up

zou3519 commented on torch/nn/utils/_expanded_weights/expanded_weights_utils.py:
```diff
@@ -0,0 +1,124 @@
+        raise RuntimeError("ExpandedWeights currently does not support a mixture of ExpandedWeight parameters "
+                           "and normal Parameters. Please file and issue with pytorch/pytorch")
+
+def sum_over_all_but_batch_and_last_n(
```
Eh, since there's tests for it here this is fine

zou3519 commented on torch/nn/utils/_expanded_weights/linear_expanded_weights.py:
```diff
@@ -0,0 +1,35 @@
+                               f"expected input of at least rank 2, got of rank {len(expanded_args[0].shape)}")
+        expanded_args = expanded_args[:-1]
+        expanded_kwargs = {'bias': expanded_args[2] if len(expanded_args) == 3 else None}
+        expanded_args_without_kwargs = expanded_args[:2]
+        output, aux_outputs = forward_helper(F.linear, expanded_args_without_kwargs, expanded_kwargs, 1)
+        ctx.args = expanded_args_without_kwargs
+        ctx.kwargs = expanded_kwargs
+        ctx.aux_outputs = aux_outputs
```
Okay the reason why I think these lines tripped me up is that they are generic and that makes them difficult to read.

Roughly speaking, here's what we're doing in the forward pass of this (and all autograd.Function for per-sample-grads):
- run f(*unexpanded_args, *unexpanded_kwargs) (or some hacked version of f, if we need intermediate values)
- save values for backward: save the required unexpanded arguments and intermediates
- return the output (as opposed to all intermediate values)

Now, the reason why this code is confusing is that it's not clear what is being saved. We are saving all the args and kwargs, but we're also saving "aux outputs", which turn out to be nothing in this case.

It would make sense for this to be generic if we planned to refactor all of the autograd.Function forward passes to look the same. Is that a good idea?

For F.linear -- there is actually no need to save bias for the backward pass, and if input does not require gradient, then there is no need to save the input! (doesn't need to happen in this PR, but those are potential optimizations). So we might not want all the autograd.Function forward passes to look similar (especially because there are already checks specific to F.linear here)

If we decide that we want this to look generic, I'd probably recommend aux_outputs be renamed to intermediates.

If we decide we don't want this to look generic, It could read better as:
```
output, = forward_helper(F.linear, expanded_args, expanded_kwargs)
ctx.unexpanded_weight = expanded_args[1]
return output
```
For e.g. group_norm this could look like:
```
output, mean, rstd = forward_helper(torch.aten.ops.native_group_norm, expanded_args, expanded_kwargs)
ctx.mean = mean
ctx.rstd = rstd
return output
```
The benefit of the non-generic form is that one doesn't have to go deep diving into the backward() part of the autograd.Function to see exactly what args, kwargs, aux_outputs are.

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights (tests in this version only test the fallback and one module, which will also call the fallback)
 - The rest is the implementation of the slow fallback + the mechanism for being able to register faster per sample grad rules. None of the faster rules are implemented here, but they are all implemented in #70141

[ghstack-poisoned]

samdow commented on torch/nn/utils/_expanded_weights/linear_expanded_weights.py:
```diff
@@ -0,0 +1,35 @@
+        bias = ctx.kwargs['bias']
+        results = []
+
+        results.append(grad_if_exists_for_input(input, lambda: grad_output.matmul(unpack_expanded_weight_or_tensor(weight))))
```
Updated. Yep this was a holdover from when we were putting all of the gradients in results so having a bunch of if/elses looked messy

samdow commented on test/test_expanded_weights.py:
```diff
@@ -0,0 +1,409 @@
+
+class TestExpandedWeightModule(TestCase):
+    def _do_test(self, module, input):
+        if sum(1 for _ in module.parameters()) == 0:  # for norms with affine=False
```
Fixed

samdow commented on torch/nn/utils/_expanded_weights/linear_expanded_weights.py:
```diff
@@ -0,0 +1,35 @@
+@implements_per_sample_grads(F.linear)
+class LinearPerSampleGrad(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, *expanded_args):
```
Updated

zou3519 reviewed (approved):
Cool!

We should probably file a follow-up issue to see if we actually duplicate the memory when using _make_subclass

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34350950).

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights
 - The rest is the implementation of the erroring fallback + the mechanism for being able to register faster per sample grad rules. Only linear is implemented here, but they are all implemented in #70141

Differential Revision: [D34350950](https://our.internmc.facebook.com/intern/diff/D34350950)

[ghstack-poisoned]

Commit:
Update on "expanded weights without fast rules"


[Design Doc for Expanded Weights](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55) <-- gives an overview of the design for Expanded Weights

Introduces the ExpandedWeights mechanism and user-facing API without any custom implemented, faster rules. 
 - User facing API is in `_stateless.py` (with documentation)
 - Testing is in test_expanded_weights
 - The rest is the implementation of the erroring fallback + the mechanism for being able to register faster per sample grad rules. Only linear is implemented here, but they are all implemented in #70141

Differential Revision: [D34350950](https://our.internmc.facebook.com/intern/diff/D34350950)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34350950).

samdow commented:
Added not user facing because we don't add release notes for prototype features. Proper tags to be added when this becomes beta
